### PR TITLE
Quick and dirty port of the OCaml pretty-print module

### PR DESCRIPTION
**Pull request by zrlk**
This implements a pretty-printing algorithm for dumping
expression trees. It's been translated more or less directly
from the corresponding F#/OCaml code, which is why it lives
in third_party. Some documentation has been reverse-engineered.

We may want to replace this in the future, but it's enough to
produce somewhat readable dumps.

### CODE REVIEW

File: anodyne/base/source_map.cc [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+      LOG(WARNING) << friendly_id << ": bad sources";
+      return false;
+    }
+    const auto& array = sources->value.GetArray();
+    for (const auto& v : array) {
+      if (!v.IsString()) {
+        LOG(WARNING) << friendly_id << ": non-string source";
+        return false;
+      }
+      sources_.emplace_back(SourceMapFile{
```
Comment (shahms) [inline anodyne/base/source_map.cc]: push_back?

File: anodyne/base/source_map_test.cc [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+#include "anodyne/base/source_map.h"
+#include "absl/strings/str_cat.h"
+#include "gtest/gtest.h"
+
+namespace anodyne {
+namespace {
+
+std::string Segment(const SourceMapSegment& s) {
+  return absl::StrCat(
+      "[", std::to_string(s.source_line), ",", std::to_string(s.source_col),
```
Comment (shahms) [inline anodyne/base/source_map_test.cc]: The to_string calls are all unnecessary

File: anodyne/base/symbol_table.h [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+namespace anodyne {
+
+/// \brief a symbol in some SymbolTable.
+using Symbol = uint32_t;
+
+/// \brief Interns strings of bytes.
+class SymbolTable {
+ public:
+  SymbolTable() {}
```
Comment (shahms) [inline anodyne/base/symbol_table.h]: explict and `=default` ?

File: anodyne/base/symbol_table.h [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    symbol_map_[std::string(text)] = static_cast<Symbol>(symbol_text_.size());
+    symbol_text_.push_back(std::string(text));
+    return symbol_text_.size() - 1;
+  }
+  /// \brief Return a symbol guaranteed to never match one of the other symbols
+  /// in this `SymbolTable`.
+  Symbol Gensym() { return gensym_++; }
+
+ private:
+  std::unordered_map<std::string, Symbol> symbol_map_;
```
Comment (shahms) [inline anodyne/base/symbol_table.h]: absl::flat_hash_map is likely a better choice (it should allow for heterogenous lookup with strings).

File: anodyne/extract/extractor.h [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ public:
+  /// \brief Attempts to identify and extract one or more compilations from
+  /// `root_path` in `file_system` and commit them to `sink`.
+  /// \param file_system the file system to read from.
+  /// \param sink where to record compilation details.
+  /// \param root_path the root path for the compilation in `file_system`.
+  /// \return true if a compilation was committed to `sink`.
+  virtual bool Extract(FileSystem* file_system, kythe::IndexWriter sink,
+                       absl::string_view root_path) = 0;
+  virtual ~Extractor() {}
```
Comment (shahms) [inline anodyne/extract/extractor.h]: prefer = default

File: anodyne/js/npm_extractor_test.cc [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+#include "kythe/cxx/common/index_writer.h"
+#include "kythe/cxx/common/json_proto.h"
+
+namespace anodyne {
+namespace {
+struct MemoryIndex {
+  /// Maps hashes to units.
+  std::map<std::string, kythe::proto::IndexedCompilation> units;
+  /// Maps hashes to file content.
+  std::map<std::string, std::string> files;
```
Comment (shahms) [inline anodyne/js/npm_extractor_test.cc]: absl::flat_hash_map?

File: anodyne/js/npm_extractor_test.cc [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+    auto it = files.find(hash);
+    if (it == files.end()) return nullptr;
+    return &it->second;
+  }
+};
+
+/// \brief Wraps a MemoryIndex.
+class MemoryIndexWriter : public kythe::IndexWriterInterface {
+ public:
+  MemoryIndexWriter(MemoryIndex* index) : memory_index_(index) {}
```
Comment (shahms) [inline anodyne/js/npm_extractor_test.cc]: explicit

File: anodyne/tools/tt.cc [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+#include <stdio.h>
+#include <string>
+
+namespace {
+
+/// \brief Unlinks a given path on destruction.
+class AutoUnlink {
+ public:
+  AutoUnlink(absl::string_view path) : path_(path) {}
```
Comment (shahms) [inline anodyne/tools/tt.cc]: explicit

File: anodyne/tools/tt_parser.cc [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+  auto pat = std::unique_ptr<TtPat>(new TtPat());
+  pat->kind = TtPat::Kind::kCtorApp;
+  pat->ident = ident;
+  pat->loc = loc;
+  for (size_t i = 0; i < pat_count; ++i) {
+    pat->children.emplace_back(
+        std::move(pat_stack_[pat_stack_.size() - pat_count + i]));
+  }
+  pat_stack_.resize(pat_stack_.size() - pat_count);
+  pat_stack_.emplace_back(std::move(pat));
```
Comment (shahms) [inline anodyne/tools/tt_parser.cc]: I suspect the emplaces here should be pushes

File: anodyne/tools/tt_parser.cc [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+}
+
+void TtParser::ApplyListPattern(const Range& loc, size_t pat_count) {
+  assert(pat_stack_.size() >= pat_count);
+  auto pat = std::unique_ptr<TtPat>(new TtPat());
```
Comment (shahms) [inline anodyne/tools/tt_parser.cc]: `absl::make_unique<TtPat>()` ? as well as the other uses of `new`

File: anodyne/tools/tt_parser.h [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+  /// Where the match happened.
+  Range loc;
+  /// Clauses in the expression.
+  std::vector<std::unique_ptr<TtClause>> clauses;
+};
+class TtParser {
+ public:
+  /// \param trace_lex Dump lexing debug information
+  /// \param trace_parse Dump parsing debug information
+  TtParser(bool trace_lex = false, bool trace_parse = false)
```
Comment (shahms) [inline anodyne/tools/tt_parser.h]: explicit; maybe make arguments into an Options struct with names?

**Comment by zrlk** [comment]
I have no clue why GitHub has chosen to show so many unrelated and/or old versions of files here, but I promise that my local enlistment has only five files changed. I'm going to try and resubmit this so it's scoped properly to its description.

**Pull request closed without merging**